src/lang/pretty_print.rs

```diff
@@ -0,0 +1,669 @@
+    }
+}
+
+impl PrettyPrint for StatementTmpl {
+    fn to_podlang_string_with_indent(&self, _indent: usize) -> String {
+        self.to_podlang_string_with_batch_context(None)
+    }
+}
+
+impl StatementTmpl {
+    fn to_podlang_string_with_batch_context(
+        &self,
+        batch_context: Option<&CustomPredicateBatch>,
+    ) -> String {
+        let mut result = String::new();
+
+        match &self.pred {
+            Predicate::Native(native_pred) => {
+                let _ = write!(&mut result, "{}", format_native_predicate(native_pred));
+            }
+            Predicate::Custom(custom_ref) => {
+                let _ = write!(&mut result, "{}", custom_ref.predicate().name);
+            }
+            Predicate::BatchSelf(index) => {
+                if let Some(batch) = batch_context {
+                    if let Some(predicate) = batch.predicates.get(*index) {
+                        let _ = write!(&mut result, "{}", predicate.name);
+                    } else {
+                        let _ = write!(&mut result, "batch_self_{}", index);
+                    }
+                } else {
+                    let _ = write!(&mut result, "batch_self_{}", index);
+                }
+            }
+        }
+
+        let _ = write!(&mut result, "(");
+        for (i, arg) in self.args.iter().enumerate() {
+            if i > 0 {
+                let _ = write!(&mut result, ", ");
+            }
+            let _ = write!(&mut result, "{}", arg.to_podlang_string());
+        }
+        let _ = write!(&mut result, ")");
+
+        result
+    }
+}
+
+impl PrettyPrint for StatementTmplArg {
+    fn to_podlang_string_with_indent(&self, _indent: usize) -> String {
+        match self {
+            StatementTmplArg::None => "none".to_string(),
+            StatementTmplArg::Literal(value) => value.to_podlang_string(),
+            StatementTmplArg::AnchoredKey(wildcard, key) => {
+                format!("?{}[{}]", wildcard.name, key)
+            }
+            StatementTmplArg::Wildcard(wildcard) => {
+                format!("?{}", wildcard.name)
+            }
+        }
+    }
+}
+
+impl PrettyPrint for CustomPredicateBatch {
+    fn to_podlang_string_with_indent(&self, indent: usize) -> String {
+        let mut result = String::new();
+
+        for (i, predicate) in self.predicates.iter().enumerate() {
+            if i > 0 {
+                let _ = write!(&mut result, "\n\n");
+            }
+            let _ = write!(
+                &mut result,
+                "{}",
+                self.format_predicate_with_context(predicate, indent)
+            );
+        }
+
+        result
+    }
+}
+
+impl CustomPredicateBatch {
+    fn format_predicate_with_context(&self, predicate: &CustomPredicate, indent: usize) -> String {
+        format_predicate_definition(predicate, indent, Some(self))
+    }
+}
+
+impl PrettyPrint for Value {
+    fn to_podlang_string_with_indent(&self, _indent: usize) -> String {
```

Comment:
There's some overlap between this implementation and `Display for TypedValue` (which is used in `Display for Value`).

The biggest changes are that in `Display` the containers are not shown fully, only their commitment is shown.

Then the syntax for `PodId, Raw, PublicKey` is different, and I think `Display for Value` would work well with the podlang syntax for those.

Perhaps we can just move this implementation to `Display for TypedValue`?  I think it's fine to show the entire contents of the containers.

Reply:
I've moved this to `TypedValue`.